### PR TITLE
Split envoyfilter into separate resources

### PR DESCRIPTION
**Pull request by DockToFuture**
Co-authored-by: johannes scheerer <johannes.scheerer@sap.com>

**How to categorize this PR?**

/area networking
/kind enhancement

**What this PR does / why we need it**:

Split envoyfilter resources from SNI and ReversedVPN into separate resources. 

**Special notes for your reviewer**:

**Release note**:

```other operator
Split `EnvoyFilter` resources from SNI and ReversedVPN into separate resources. 
```

### CODE REVIEW

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go [hunk 1]
```diff
@@ -567,6 +582,183 @@ func (v *vpnSeedServer) Deploy(ctx context.Context) error {
...
+					UID:                v.namespaceUID,
+					Controller:         pointer.BoolPtr(false),
+					BlockOwnerDeletion: pointer.BoolPtr(false),
+				},
+			},
+		}
+		envoyFilter.Spec.WorkloadSelector = &istionetworkingv1alpha3.WorkloadSelector{
+			Labels: v.istioIngressGateway.Labels,
+		}
+		envoyFilter.Spec.ConfigPatches = []*istionetworkingv1alpha3.EnvoyFilter_EnvoyConfigObjectPatch{
```
Comment (rfranzke) [inline pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go:604]: I'm getting dizzy when looking at the nesting here. That's exactly the reason why I went with the YAML approach back as part of #3983. I don't insist on changing this again, but do you want to reconsider this decision? :)

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go [hunk 2]
```diff
@@ -567,6 +582,183 @@ func (v *vpnSeedServer) Deploy(ctx context.Context) error {
...
+																									StringValue: v.namespace,
+																								},
+																							},
+																							"domains": &protobuftypes.Value{
+																								Kind: &protobuftypes.Value_ListValue{
+																									ListValue: &protobuftypes.ListValue{
+																										Values: []*protobuftypes.Value{
+																											{
+																												Kind: &protobuftypes.Value_StringValue{
+																													StringValue: *v.kubeAPIServerHost + ":8132",
```
Comment (rfranzke) [inline pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: Where is `8132` coming from? I see it multiple times, maybe we should have a constant somewhere?

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go [hunk 3]
```diff
@@ -567,6 +582,183 @@ func (v *vpnSeedServer) Deploy(ctx context.Context) error {
 		return err
 	}
 
+	if _, err := controllerutil.CreateOrUpdate(ctx, v.client, envoyFilter, func() error {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: See #4227, maybe you can go with `GetAndCreateOrMergePatch` for this new code right away.

**Review by rfranzke — COMMENTED** [review]
Thanks for the PR, I like it a lot! :)

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go (hunk 1 above)
Comment (DockToFuture) [reply pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go:604]: We did it the first time as well with the YAML approach however we had problems to fix the tests and then we decided to go this way and do it like in the rest of the file with the `CreateOrUpdate` or call from controllerruntime.

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go (hunk 2 above)
Comment (DockToFuture) [reply pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: Introduced a constant.

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go (hunk 3 above)
Comment (DockToFuture) [reply pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: Used it for the new code and adapted the tests accordingly.

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go [hunk 4]
```diff
@@ -49,6 +53,8 @@ const (
 	DeploymentName = v1beta1constants.DeploymentNameVPNSeedServer
 	// ServiceName is the name of the vpn seed server service running internally on the control plane in seed.
 	ServiceName = DeploymentName
+	// GatewayPort is the port exposed by the istio ingress gateway for the vpn seed service
+	GatewayPort = 8132
```
Comment (rfranzke) [inline pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: Shouldn't this rather be in the `istio` package and not here?

Comment (DockToFuture) [reply pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: Why? This port is some vpn-seed-server specific configuration.

Comment (rfranzke) [reply pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: Is it? OK, then the doc string was misleading/confusing (at least for me, looking at all of this without being into the details).

**Review by rfranzke — COMMENTED** [review]
What about https://github.com/gardener/gardener/blob/eb2f2907925de8ae1e90cceeeafa22c4b6bfc524/charts/istio/istio-ingress/templates/envoy-filter.yaml#L202-L231? Doesn't this also belong to the `vpnseedserver` component?

Also, the new port constant should probably be used here: https://github.com/gardener/gardener/blob/a1c98458113d13b290cfc59404091f3a17fdf70d/pkg/operation/seed/seed.go#L723?

**Comment by DockToFuture** [comment]
> What about
> 
> https://github.com/gardener/gardener/blob/eb2f2907925de8ae1e90cceeeafa22c4b6bfc524/charts/istio/istio-ingress/templates/envoy-filter.yaml#L202-L231
> 
> ? Doesn't this also belong to the `vpnseedserver` component?
> Also, the new port constant should probably be used here:
> 
> https://github.com/gardener/gardener/blob/a1c98458113d13b290cfc59404091f3a17fdf70d/pkg/operation/seed/seed.go#L723
> 
> ?

I think it is in there because it is also used by the konnectivity tunnel, which uses 8132 port.

**Comment by rfranzke** [comment]
@DockToFuture OK, for https://github.com/gardener/gardener/blob/a1c98458113d13b290cfc59404091f3a17fdf70d/pkg/operation/seed/seed.go#L723 this seems to be correct, but https://github.com/gardener/gardener/blob/eb2f2907925de8ae1e90cceeeafa22c4b6bfc524/charts/istio/istio-ingress/templates/envoy-filter.yaml#L202-L231 was initially contributed with `ReversedVPN` feature gate PR (#3812)?

**Comment by ScheererJ** [comment]
> @DockToFuture OK, for
> 
> https://github.com/gardener/gardener/blob/a1c98458113d13b290cfc59404091f3a17fdf70d/pkg/operation/seed/seed.go#L723
> 
> this seems to be correct, but
> https://github.com/gardener/gardener/blob/eb2f2907925de8ae1e90cceeeafa22c4b6bfc524/charts/istio/istio-ingress/templates/envoy-filter.yaml#L202-L231
> 
> was initially contributed with `ReversedVPN` feature gate PR (#3812)?

The http-connect-listener is deployed once per seed, i.e. in the seed bootstrap, while the other `ReverseVPN` components are deployed per shoot. I think that it fits rather nicely next to the other generic `EnvoyFilter`s. The other ones are also not moved to the sni components, but stay with istio-ingress.

**Comment by rfranzke** [comment]
I'm fine for now (no need to change this PR), but I still think this should be improved. With #2754 we aim to simplify the complexity and distribution of code related to a single component so that the entire data flow is easier to understand, maintain and develop. Hence, I propose to move the `EnvoyFilter` that is only deployed once per seed into the `vpnseedserver` component by introducing something similar to https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/component/clusterautoscaler/bootstrap.go.
The same should probably be done for the `kube-apiserver` related resources.

As said, no need to do it now, maybe you have time in the next weeks to do it, maybe we can do it ourselves, let's see. Just wanted to point it out. :)

File: pkg/operation/botanist/component/istio/type.go [hunk 5]
```diff
@@ -16,4 +16,6 @@ package istio
 
 const (
 	istioReleaseName = "istio"
+	// GatewayPort is the port exposed by the istio ingress gateway
+	GatewayPort = 8132
```
Comment (rfranzke) [inline pkg/operation/botanist/component/istio/type.go:20]: I don't quite get why you have moved this now. Didn't you just argue in https://github.com/gardener/gardener/pull/4242#discussion_r656879036 that the port is related to `vpnseedserver`? My comment was only about improving the doc string to make it more clear.

Anyways, I'll approve the PR now.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**